### DISCUSSION TITLE
Tasks: Scripting support + Built-In variables

### DISCUSSION THREAD

**Discussion by erohtar** [adirh3/Fluent-Search#613] [General]
**Is your feature request related to a problem? Please describe.**
Tasks functionality has a potential to become super-powerful in future, but it is very limited right now. 
I'm proposing to add proper scripting support to Tasks, as in multiple commands in a script that execute one after the other, not just a single nested C# command that can be put in a single variable/output field.
The scripts would also have access to read **FS provided built-in variables** (like portable path, current task/context etc).
**Scripting support wouldn't just mean launching a script**, but also waiting for it to finish and use its output in next Task element.

I'm not asking for FS adding its own scripting language from scratch (a monumental task by any measure) but **FS could/should open doors to an existing scripting language on Windows** (my favorite and recommendation would be AutoHotkey, but there are other options) - where a task could launch a script so that:
- Script is able to interact with FS, access built-in variables
- FS is able to access variables that script makes available
- FS is able to wait for script to finish and use/display output in own UI

**Plugins vs Scripts**
While you might contend that what I'm talking about is a domain for plugins, but scripts have a lot of benefits that plugins don't:
- Not everyone is able to write plugins
- Learning to write a script has a much flatter learning curve than learning to write a plugin (in any language)
- All popular scripting languages already have excellent documentation available, not so for FS plugins right now
- A plugin once authored, can not be tweaked by users for slight customizations unless author added support for that, but script can be
- Adding support for existing popular scripting language could potentially bring in a whole new set of users to FS

**Comment by adirh3** [comment]
@erohtar thanks for the detailed request, and sorry for not replying for a while, I have been trying to focus on basics for the last few months.
We have talked about this in depth in Discord, but writing it here for the sake of sharing with everyone.

After looking for a while I decided to go with PS since it's most commonly used scripting language in Windows.
The current situation with Tasks, it supports scripting via PowerShell or C# scripts inside Tasks.

What you can currently do (0.9.91.x) -
- Script can interact with all variables, using `$myVar` in PS or `myVar` in C#
- You can output variables using `Write-Host ##fsvar varname value` in PS or `return myVar` in C#
- You can configure PS scripts to wait for completion and/or show the PS window
- Script errors will show in the Task window **(new)**

What's not available (as date of writing this) -
- No extensive documentation for Tasks
- No support for other scripting via files, but you can still run any file through PS

What's planned -
- Support for Python script as first-class citizen inside FS, with all features as PS and C# script
- Support for triggering any script (through file) and receiving results through a generic output (e.g. stdout)

I hope this gives a clearer picture, please let me know what you think.
Also, changing this to a discussion as I think it's might bring more voices :)

**Comment by erohtar** [comment]
Thanks for replying. Though I don't have any experience with those scripting languages, it's still good to have support for them. And they're all very popular so it should be easy enough to look up and refer. Problem would still be the lack of documentation on Tasks, though I'm sure you'll get to that eventually.

**Reply by riotrah** [reply]
If it helps, PS has some interesting idiosyncrasies in syntax, but it's really really easy to pick up and easy to REPL-driven dev using PS shell itself
